File: src/SolidMaterial/StrengthModel.cc
//---------------------------------Spheral++----------------------------------//
// StrengthModel -- The interface base class for strength models.
//
// Created by JMO, Wed Sep 8 15:18:44 2004
//----------------------------------------------------------------------------//
#include "StrengthModel.hh"
#include "Field/Field.hh"

namespace Spheral {

//------------------------------------------------------------------------------
// Constructor
//------------------------------------------------------------------------------
template<typename Dimension>
StrengthModel<Dimension>::
StrengthModel() {
}

//------------------------------------------------------------------------------
// Destructor
//------------------------------------------------------------------------------
template<typename Dimension>
StrengthModel<Dimension>::
~StrengthModel() {
}

//------------------------------------------------------------------------------
// Sound speed
//------------------------------------------------------------------------------
template<typename Dimension>
void
StrengthModel<Dimension>::
soundSpeed(Field<Dimension, Scalar>& soundSpeed,
           const Field<Dimension, Scalar>& /*density*/,
           const Field<Dimension, Scalar>& /*specificThermalEnergy*/,
           const Field<Dimension, Scalar>& /*pressure*/,
           const Field<Dimension, Scalar>& fluidSoundSpeed,
           const Field<Dimension, SymTensor>& /*damage*/) const {
  soundSpeed = fluidSoundSpeed;
}

//------------------------------------------------------------------------------
// Bulk modulus
//------------------------------------------------------------------------------
template<typename Dimension>
void
StrengthModel<Dimension>::
bulkModulus(Field<Dimension, Scalar>& /*bulkModulus*/,
            const Field<Dimension, Scalar>& /*density*/,
            const Field<Dimension, Scalar>& /*specificThermalEnergy*/) const {
  VERIFY2(false,
          "StrengthModel::bulkModulus called");
}

//------------------------------------------------------------------------------
// Melt energy
//------------------------------------------------------------------------------
template<typename Dimension>
void
StrengthModel<Dimension>::
meltSpecificEnergy(Field<Dimension, Scalar>& meltSpecificEnergy,
                   const Field<Dimension, Scalar>& /*density*/,
                   const Field<Dimension, Scalar>& /*specficThermalEnergy*/) const {
  meltSpecificEnergy = 0.0;
}

//------------------------------------------------------------------------------
// Cold energy
//------------------------------------------------------------------------------
template<typename Dimension>
void
StrengthModel<Dimension>::
coldSpecificEnergy(Field<Dimension, Scalar>& coldSpecificEnergy,
                   const Field<Dimension, Scalar>& /*density*/,
                   const Field<Dimension, Scalar>& /*specficThermalEnergy*/) const {
  coldSpecificEnergy = 0.0;
}

}

File: src/Pybind11Wraps/SolidMaterial/StrengthModel.py
#-------------------------------------------------------------------------------
# StrengthModel abstract class
#-------------------------------------------------------------------------------
from PYB11Generator import *
from StrengthModelAbstractMethods import *

@PYB11template("Dimension")
@PYB11module("SpheralSolidMaterial")
class StrengthModel:
    "Abstract base for strength models"

    PYB11typedefs = """
    using Scalar = typename %(Dimension)s::Scalar;
    using SymTensor = typename %(Dimension)s::SymTensor;
    using ScalarField = Field<%(Dimension)s, Scalar>;
"""

    #...........................................................................
    # Constructors
    def pyinit(self):
        "Default constructor"

    #...........................................................................
    # Virtual methods
    @PYB11virtual
    @PYB11const
    def providesSoundSpeed(self):
        return "bool"

    @PYB11virtual
    @PYB11const
    def providesBulkModulus(self):
        return "bool"

    @PYB11virtual
    @PYB11const
    def soundSpeed(self,
                   soundSpeed = "Field<%(Dimension)s, Scalar>&",
                   density = "const Field<%(Dimension)s, Scalar>&",
                   specificThermalEnergy = "const Field<%(Dimension)s, Scalar>&",
                   pressure = "const Field<%(Dimension)s, Scalar>&",
                   fluidSoundSpeed = "const Field<%(Dimension)s, Scalar>&",
                   damage = "const Field<%(Dimension)s, SymTensor>&"):
        return "void"

    @PYB11virtual
    @PYB11const
    def bulkModulus(self,
                    bulkModulus = "Field<%(Dimension)s, Scalar>&",
                    massDensity = "const Field<%(Dimension)s, Scalar>&",
                    specificThermalEnergy = "const Field<%(Dimension)s, Scalar>&"):
        return "void"

    @PYB11virtual
    @PYB11const
    def meltSpecificEnergy(self,
                           meltSpecificEnergy = "Field<%(Dimension)s, Scalar>&",
                           density = "const Field<%(Dimension)s, Scalar>&",
                           specficThermalEnergy = "const Field<%(Dimension)s, Scalar>&"):
        return "void"

    @PYB11virtual
    @PYB11const
    def coldSpecificEnergy(self,
                           coldSpecificEnergy = "Field<%(Dimension)s, Scalar>&",
                           density = "const Field<%(Dimension)s, Scalar>&",
                           specficThermalEnergy = "const Field<%(Dimension)s, Scalar>&"):
        return "void"

#-------------------------------------------------------------------------------
# Inject abstract interface
#-------------------------------------------------------------------------------
PYB11inject(StrengthModelAbstractMethods, StrengthModel, virtual=False, pure_virtual=True)
